IOSExercises/Views/DetailsViewController.m

```diff
@@ -0,0 +1,341 @@
+        /*
+         * Display as action sheet
+         * Currently logs layout constraints conflict error
+         * Looks like a bug from Apple
+         * See https://github.com/lionheart/openradar-mirror/issues/21120
+         */
+        [self displayColorPickerActionSheet];
+    } else {
+        [self popoverColorPicker];
+    }
+}
+
+- (void)showDatePicker:(UITapGestureRecognizer *)recognizer {
+    [self popoverDateTimePicker];
+}
+
+- (IBAction)greetIfBirthday:(id)sender {
+    NSDateFormatter *dateFormatter = [self birthdayFormatter];
+
+    NSDate *birthdate = [dateFormatter dateFromString:self.birthdayLabel.text];
+
+    NSDateComponents *bdayComponent = [[NSCalendar currentCalendar] components:NSCalendarUnitDay | NSCalendarUnitMonth
+                                                                      fromDate:birthdate];
+    NSDateComponents *todayComponent = [[NSCalendar currentCalendar] components:NSCalendarUnitDay | NSCalendarUnitMonth
+                                                                       fromDate:[NSDate date]];
+
+    NSString *greeting;
+
+    // Currently comparing day and month date components instead of inSameDayAsDate
+    if (birthdate
+        && bdayComponent.day == todayComponent.day
+        && bdayComponent.month == todayComponent.month) {
+        greeting = @"Happy birthday!";
+    } else {
+        greeting = @"It's not your birthday";
+    }
+    [self popoverWithGreeting:greeting];
+}
+
+- (IBAction)chooseColorRed:(id)sender {
+    [self.colorLabel setText:@"Red"];
+}
+
+- (IBAction)chooseColorBlue:(id)sender {
+    [self.colorLabel setText:@"Blue"];
+}
+
+- (IBAction)chooseColorBlack:(id)sender {
+    [self.colorLabel setText:@"Black"];
+}
+
+- (IBAction)saveBirthday:(id)sender {
+    if (self.datePicker) {
+        NSLog(@"Saving birthday");
+        [self setBirthdayWithDate:[self.datePicker date]];
+    }
+    NSLog(@"DetailsViewController: Birthday was not saved");
+
+    [self.popoverViewController dismissViewControllerAnimated:YES
+                                                   completion:nil];
+    [self promptErrorWithMessage:@"Changes have not been saved"];
+}
+
+#pragma mark - Popover View Controller Methods
+
+- (void)popoverDateTimePicker{
+    if (!self.popoverViewController) {
+        self.popoverViewController = [[UIViewController alloc] init];
+    }
+    
+    self.popoverViewController.view = [[UIView alloc] init];
+    [self.popoverViewController setPreferredContentSize:CGSizeMake(250.f, 270.f)];
+
+    //[self setBirthday];
+    if (!self.datePicker) {
+        self.datePicker = [[UIDatePicker alloc] init];
+    }
+    [self.datePicker setTranslatesAutoresizingMaskIntoConstraints:NO];
+    [self.datePicker setDatePickerMode:UIDatePickerModeDate];
+    [self.popoverViewController.view addSubview:self.datePicker];
+
+    UIButton *saveButton = [UIButton buttonWithType:UIButtonTypeSystem];
```

Comment:
Refactor method when possible (and just group the helper methods under a #pragma mark, for example)